First version of games

pjkedl opened the pull request:
These files included the first version of the games Tic tac toe and Bulls and Cows.
Work TO BE:
Better play board for tic tac toe. How?

Waiting for your review and suggestions. 
Thank you
Pavel

pjkedl committed:
- Passwords dictionary made more readable
- abbreviated variable usr changed to user
- iterated variable i changed to attempt
- abbreviated variable pwd changed to password
- login part infinite loop corrected + 3rd wrong attempt exits program
- abbreviated variable sel changed to choice

pjkedl committed:
- abbreviated variables corrected

pjkedl committed:
- Readme for project 1 Text analyzer from engeto python academy
- Added task description

pjkedl committed:
- first version of function show_board

pjkedl committed:
- show_board function
- check_winner function
- ask_for_move function

pjkedl committed:
- Tic tac toe readme

pjkedl committed:
- Tic tac toe testing version 0.99
- not tested every possibility, game is playable
- Need better show_board() functionality

pjkedl committed:
- Function check_board() added - check if there is move possible

pjkedl committed:
- Bulls and Cows readme file

pjkedl committed:
- Bulls and Cows game working file

pjkedl committed:
- Bulls and Cows game test with several issues

pjkedl committed:
- Bulls and Cows game version 1.0

Artimi commented on Project3_Bulls_and_Cows/Game.py:
```diff
@@ -0,0 +1,67 @@
+
+
+def number_generator():
+    number = str(random.randint(1000,9999))
```
This prevents numbers like digits like `0123` to be generated. I think it should be there.

Also, this can generate a number like `1111` which contradicts the rules:
> The digits must be all different.

pjkedl replied:
Added new version. Now it should work.

Artimi commented on Project3_Bulls_and_Cows/Game.py:
```diff
@@ -0,0 +1,67 @@
+    game_header()
+    computer_list = number_generator()
+    while True:
+        bull_cow(counter, ask_for_number(), computer_list)
```
You are putting `ask_for_number()` as the second argument but the function has `computer_list` as the second argument. You have swapped the meaning of the numbers.

Also, try to think about better names than `computer_list` and `my_list`. What about `secret` and `guess`?

pjkedl replied:
Code updated. Variables changed and lists are now in right order.

Artimi commented on Project3_Bulls_and_Cows/Game.py:
```diff
@@ -0,0 +1,67 @@
+        exit()
+    elif bull == 4 and counter > 25:
+        print(f"Correct, you've guessed the right number in {counter} guesses! You can do better.")
+        exit()
```
I think that you should not call `exit()` elsewhere than in `main()`. It might be sometimes confusing when the program does not have a chance to do some cleaning. I would rather return `False` in case a user did not guess the secret correctly and `True` otherwise.

The code here seems quite duplicated. What about using helper function that will evaluate the performance:
```python
def evaluate_performance(guesses):
  if counter <= 7:
      return "That's excellent performance"
  elif counter <= 15:
    return "That's very good performance"
  elif counter <= 25:
    return "That's not very good performance"
  else:
    return "You can do better"
...
elif bull == 4:
  print(f"Correct, you've guessed the right number in {counter} guesses! {evaluate_performance(counter)}"
  return True
```
This places information about your performance to a single place and you can easily add or remove a message.

Also note that you are missing some number when using `range` in such way. `range(8, 15)` does not contain 15 because the second parameter is exclusive.

Artimi commented on Project3_Bulls_and_Cows/Game.py:
```diff
@@ -0,0 +1,67 @@
+    counter = 1
+    game_header()
+    computer_list = number_generator()
+    while True:
```
If you want to make it fancy you could use https://docs.python.org/3/library/itertools.html#itertools.count
```python
import itertools
...
for counter in itertools.count(1):
  ...
```
And you don't have to worry about `counter` by yourself.

Artimi commented on Project3_Bulls_and_Cows/Game.py:
```diff
@@ -0,0 +1,67 @@
+    while True:
+        bull_cow(counter, ask_for_number(), computer_list)
+        counter += 1
+        print(counter)
```
The counter is not printed in the task assignment.

Artimi commented on Project2_TicTacToe/game.py:
```diff
@@ -0,0 +1,98 @@
+def show_board(my_list):
+    N = 3
+    print('=' * 10)
+    print(f'\n'.join([' |'.join(i) for i in zip(*[iter(my_list)] * N)]))
```
Cool trick with `zip` and `iter`. However, I would write a comment to the line above it to describe what it actually does because it can be confusing for someone who does not know the construct (or even you after some time).

Artimi replied:
You could fix the problem with different width of gaming board by having space `' '` as the empty symbol in the board and then putting two spaces around the pipe `|`
```python
print('\n'.join([' | '.join(i) for i in zip(*[iter(my_list)] * N)]))
```

Artimi commented on Project2_TicTacToe/game.py:
```diff
@@ -0,0 +1,98 @@
+                continue
+            else:
+                return index, char
+                break
```
This is a typical code screaming for duplication. These few lines are the same but one line with char assignment. What about:
```python
if turn % 2 == 0:
  char = 'O'
else:
  char = 'X' 
while True:
  try:
    index = int(input(f'Player {char} | Please enter your move number: '))
  except ValueError:
     print('Wrong input')
     continue
  else:
    return index, char
  ```
Also, `break` after `return` does not make sense. The code will never reach the `break` statement.

Artimi replied:
As I'm looking at it now I would remove `turn` from `ask_for_move` altogether and used `char` instead (which you can precompute somewhere). Then your function `ask_for_move` doesn't have to know about turns and that it should be altering chars. Now `ask_for_move` handles two responsibilities: getting the correct move **and** returning next `char`. A function should have only one responsibility https://en.wikipedia.org/wiki/Single_responsibility_principle

Artimi commented on Project2_TicTacToe/game.py:
```diff
@@ -0,0 +1,98 @@
+        return False
+    else:
+        return True
+def check_board(board):
```
There should be a blank line between functions

Artimi commented on Project2_TicTacToe/game.py:
```diff
@@ -0,0 +1,98 @@
+    if dataset[index-1] != '':
+        return False
+    else:
+        return True
```
This can be simplified to
```python
def check_move(index):
  return dataset[index-1] == ''
```

Artimi commented on Project2_TicTacToe/game.py:
```diff
@@ -0,0 +1,98 @@
+def check_board(board):
+    if not '' in board:
+        print('No more moves possible. Game has no winner. GAME OVER')
+        exit()
```
Don't exit from the called function.

Artimi commented on Project2_TicTacToe/game.py:
```diff
@@ -0,0 +1,98 @@
+    if not '' in board:
+        print('No more moves possible. Game has no winner. GAME OVER')
+        exit()
+def check_winner(board):
```
Blank line between functions.

Artimi commented on Project2_TicTacToe/game.py:
```diff
@@ -0,0 +1,98 @@
+            (board[2] == 'O' and board[4] == 'O' and board[6] == 'O') or
+            (board[0] == 'O' and board[3] == 'O' and board[6] == 'O') or
+            (board[1] == 'O' and board[4] == 'O' and board[7] == 'O') or
+            (board[2] == 'O' and board[5] == 'O' and board[8] == 'O')):
```
You are duplicating information about which configurations are winning for both players. Either you could put the code to
```python
for char in ('X', 'O'):
  ...
```
Or I suggest to prepare a better list of winning moves:
```python
WINNING_CONFIGURATIONS = (
  (0, 1, 2),
  (3, 4, 5),
  ...
  (2, 5, 8)
)
...
def check_winner(board):
  for x, y, z in WINNING_CONFIGURATIONS:
    if (board[x] == board[y] == board[z] != ''):
      return True
  return False
```
It's important to check that the field is not empty with `!= ''` or else it would claim it winning even for empty board.
It is just a suggestion, it's up to you whether you want to use it or not.

Artimi commented on Project2_TicTacToe/game.py:
```diff
@@ -0,0 +1,98 @@
+    while True:
+        check_board(dataset)
+        index, char = ask_for_move(turn)
+        if check_move(index) == False:
```
Rather use
```python
if not check_move(index):
```

Artimi replied:
Maybe than having `check_move` here. You could put it to `ask_for_move` and then you would stay in the function until a user enters the correct move. What do you think?

Artimi commented on Project2_TicTacToe/game.py:
```diff
@@ -0,0 +1,98 @@
+            move(index, char)
+            show_board(dataset)
+            turn += 1
+        if check_winner(dataset) == True:
```
```python
if check_winner(dataset):
```
Is sufficient

Artimi commented on Project2_TicTacToe/readme.txt:
```diff
@@ -0,0 +1 @@
+This project is a game of Tic Tac Toe 3x3 size
```
You could add the assignment here too.

Artimi commented on Project2_TicTacToe/show_board.py:
```diff
@@ -0,0 +1,5 @@
+    N = 3
+    print('=' * 10)
+    print('\n'.join([' |'.join(i) for i in zip(*[iter(my_list)] * N)]))
+    print('=' * 10)
```
This shouldn't be here, right? It is just some testing code.

pjkedl committed:
- new version of secret number generator

pjkedl committed:
- new version of secret number generator

pjkedl committed:
- new function evaluate_performance added
- exit() code only in main function
- changed uncertain variable names computer_list -> secret
  and my_list -> guess

pjkedl committed:
- itertools added for counter variable

Artimi commented on Project3_Bulls_and_Cows/Game.py:
```diff
@@ -0,0 +1,75 @@
+    generated_number = [int(x) for x in range(10)]
+    secret = []
+
+    for i in generated_number:
```
What about using condition
```python
while len(secret) < 4:
```
Then you will express the intent and you don't even have to break from the loop. You don't use `i` anyway. And also you should not remove elements from the collection you are looping over.

Artimi commented on Project3_Bulls_and_Cows/Game.py:
```diff
@@ -0,0 +1,75 @@
+        except ValueError:
+            print('Please enter number only.')
+            continue
+        if a not in range(1000, 9999):
```
This still checks that a user enter a number larger than `1000` but secret can now start with `0`. Good luck with guessing `0123` because you won't be able to input the number.

Artimi commented on Project3_Bulls_and_Cows/Game.py:
```diff
@@ -0,0 +1,75 @@
+        else:
+            guess = [int(x) for x in str(a)]
+            break
+    return guess
```
You don't have to break from the loop and just return
```python
return [int(x) for x in str(a)]
```

Artimi commented on Project3_Bulls_and_Cows/Game.py:
```diff
@@ -0,0 +1,75 @@
+    game_header()
+    secret = number_generator()
+    for counter in itertools.count(1):
+        if bull_cow(counter, ask_for_number(), secret) == True:
```
There is no need to check equality with `True`
```python
if bull_cow(counter, ask_for_number(), secret):
  ...
```
This has the same meaning but is shorter and more used.

Artimi commented on Project3_Bulls_and_Cows/Game.py:
```diff
@@ -0,0 +1,75 @@
+    for counter in itertools.count(1):
+        if bull_cow(counter, ask_for_number(), secret) == True:
+            break
+    exit()
```
Why do you call `exit` here? The code will exit anyway because it will jump out of `main` function.

pjkedl committed:
- Project election scraper first version

pjkedl committed:
- Added readme file

Artimi commented on Project4_election_scraper/Project4_election_scraper.py:
```diff
@@ -0,0 +1,101 @@
+import requests
```
You should add a new, unrelated code to already reviewed (before merge) pull requests. It is best to create one pull request for one change only.

Artimi commented on Project4_election_scraper/Project4_election_scraper.py:
```diff
@@ -0,0 +1,101 @@
+
+domain = 'https://volby.cz/pls/ps2017nss/'
+district_links = []
+header_url = 'https://www.volby.cz/pls/ps2017nss/ps2?xjazyk=CZ'
```
You should write constants at the top of the file and in UPPER_CASE

Artimi commented on Project4_election_scraper/Project4_election_scraper.py:
```diff
@@ -0,0 +1,101 @@
+header_url = 'https://www.volby.cz/pls/ps2017nss/ps2?xjazyk=CZ'
+
+def municipality_links():
+    soup = load_soup('https://volby.cz/pls/ps2017nss/ps3?xjazyk=CZ')
```
Rather than having the URL here you could store it also to a constant at the file beginning.

Artimi commented on Project4_election_scraper/Project4_election_scraper.py:
```diff
@@ -0,0 +1,101 @@
+    print('Loading soup https://volby.cz/pls/ps2017nss/ps3?xjazyk=CZ')
+    for href in soup.find_all('a', href=True):
+        if 'ps32' in href['href']:
+            district_links.append(domain+href['href'])
```
Missing spaces around operator `+`

Artimi commented on Project4_election_scraper/Project4_election_scraper.py:
```diff
@@ -0,0 +1,101 @@
+    municipality_links = []
+    for link in district_links:
+        response = requests.get(link)
+        soup = BS(response.content, 'html.parser')
```
Why don't you use `load_soup` function here?

Artimi commented on Project4_election_scraper/Project4_election_scraper.py:
```diff
@@ -0,0 +1,101 @@
+    for link in district_links:
+        response = requests.get(link)
+        soup = BS(response.content, 'html.parser')
+        for td in soup.find_all('td',{'class':'cislo'}):
```
Missing space after `,` and `:`

Artimi replied:
Why don't you use `class_` argument also here?

Artimi commented on Project4_election_scraper/Project4_election_scraper.py:
```diff
@@ -0,0 +1,101 @@
+                city_link = (domain+municipality['href'])
+                municipality_links.append([city, city_link])
+    return municipality_links
+
```
Top-level functions should be surrounded by 2 blank lines https://pep8.org/#blank-lines

Artimi commented on Project4_election_scraper/Project4_election_scraper.py:
```diff
@@ -0,0 +1,101 @@
+        party_number = element.find('td', class_="cislo", headers="t1sa1 t1sb1")
+        if party_number != None:
+            row_dict[party_number.get_text()] = cell.get_text().replace('\xa0','')
+            # row.append(cell.get_text().replace('\xa0',''))
```
Why is this code here? It is a dead code. Either delete it or explain the reason or you will forget why you have it here.

Artimi commented on Project4_election_scraper/Project4_election_scraper.py:
```diff
@@ -0,0 +1,101 @@
+    for header in soup.find_all(class_="topline", id="publikace"):   #scraping city name
+        city = header.find_all('h3')[-1].get_text()
+        city = city.replace('\n', '')
+        city = city.replace('Obec: ', '')
```
You are doing this in a for loop but don't use the `city` variable. You only use the last one on line 65. Why?

Artimi commented on Project4_election_scraper/Project4_election_scraper.py:
```diff
@@ -0,0 +1,101 @@
+        party_number2 = element2.find('td', class_="cislo", headers="t2sa1 t2sb1")
+        if party_number2 != None:
+            row_dict[party_number2.get_text()] = cell2.get_text().replace('\xa0','')
+            # row.append(cell2.get_text().replace('\xa0',''))
```
These two blocks of code look very similar. Think of a way to put this into a function with some arguments so you can reuse the code.

Artimi commented on Project4_election_scraper/Project4_election_scraper.py:
```diff
@@ -0,0 +1,101 @@
+        writer = csv.writer(file, delimiter=';', quotechar=' ')
+        for row in data:
+            writer.writerow(row)
+def main():
```
Two blank lines between functions

Artimi commented on Project4_election_scraper/Project4_election_scraper.py:
```diff
@@ -0,0 +1,101 @@
+            writer.writerow(row)
+def main():
+    all_data = []
+    all_data.append(scrape_data_header(header_url))
```
Why not write this on a single line
```python
all_data = [scrape_data_header(header_url)]
```

Artimi commented on Project4_election_scraper/Project4_election_scraper.py:
```diff
@@ -0,0 +1,101 @@
+    for element2 in election_results_table2.find_all('tr'):
+        cell = element2.find('td', headers="t2sa1 t2sb2")
+        if cell != None:
+            header_row.append(cell.get_text())
```
These are also two similar blocks of code. Think about putting it in a function and calling it twice.

Artimi commented on Project4_election_scraper/Project4_election_scraper.py:
```diff
@@ -0,0 +1,101 @@
+        if party_number2 != None:
+            row_dict[party_number2.get_text()] = cell2.get_text().replace('\xa0','')
+            # row.append(cell2.get_text().replace('\xa0',''))
+    for i in range(1, 32):
```
The number of values is hardcoded, when you will use this code for another election you will need to change it here. What about getting this from header and passing it as argument to the function?

Artimi commented on Project4_election_scraper/Project4_election_scraper.py:
```diff
@@ -0,0 +1,101 @@
+        if i in row_dict:
+             row.append(row_dict[i])
+        else:
+            row.append('0')
```
Instead of this conditition you could write
```python
row.append(row_dict.get(str(i))
```

Artimi commented on Project4_election_scraper/Project4_election_scraper.py:
```diff
@@ -0,0 +1,101 @@
+    return BS(response.text, 'html.parser')
+
+domain = 'https://volby.cz/pls/ps2017nss/'
+district_links = []
```
Why is `district_links` global? You are using it only in `municipality_links` so define it just there.

Artimi commented on Project4_election_scraper/Project4_election_scraper.py:
```diff
@@ -0,0 +1,101 @@
+    envelopes = soup.find(id='ps311_t1').find('td', class_="cislo", headers="sa3").get_text()
+    valid = soup.find(id='ps311_t1').find('td', class_="cislo", headers="sa6").get_text()
+    row_dict = {}
+    row = [code, city, str(registered).replace('\xa0',''), str(envelopes).replace('\xa0',''), str(valid).replace('\xa0','')]
```
`str(value).replace('\xa0', '')` is repeated a lot in the code. What about putting it in a function?
```python
def clean(value):
    return str(value).replace('\xa0', '')
```

Artimi commented on Project4_election_scraper/README:
```diff
@@ -0,0 +1,14 @@
+Link where the list of the municipalities is located
+name of the csv file without the suffix .csv
+Output
+The header of the csv file should have the following: municipality code, name of location, voters in the list, distributed voting envelopes, valid votes, candidate parties. Each row of the csv file will then represent individual municipalities. So an example of Hradec Králové would look like this:
```
Great description. Could you also add an example of running this script (just the script with command line arguments) so a user can try it right away?

pjkedl committed:
Project election scraper:
- constants corrected to upper_case and moved to the beginning of the code
- code formatted to standards
- added new function clean() for cleaning scraped text values
- variable district_links is now local in function municipality_links()
- corrected class_ argument
- dead code deleted
